pysoa/common/logging.py

```diff
@@ -199,3 +202,179 @@ def _copy_and_censor_dict(cls, d):
     @classmethod
     def _copy_and_censor_iterable(cls, i, should_censor_strings):
         return type(i)(cls._copy_and_censor_unknown_value(v, should_censor_strings) for v in i)
+
+
+IP_MTU_DISCOVER = 10  # Position of the IP Path MTU Discovery flag in request packets
+IP_MTU_DISCOVER_DO = 2  # "Don't fragment" value of the IP Path MTU Discovery flag in request packets
+WORST_CASE_MTU_IP = 576
+MTU_ATTEMPTS = 65535, 2000, 1500, 1280
+DATAGRAM_HEADER_LENGTH_IN_BYTES = 28
+
+
+def _discover_minimum_mtu_to_target(address, port):
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    s.connect((address, port))
+    s.setsockopt(socket.IPPROTO_IP, IP_MTU_DISCOVER, IP_MTU_DISCOVER_DO)
+    for attempt in MTU_ATTEMPTS:
+        try:
+            s.send(b'#' * (attempt - DATAGRAM_HEADER_LENGTH_IN_BYTES))
+        except socket.error as e:
+            if 'too long' not in e.strerror:
+                # We can't rely on the error code, because it varies from platform to platform, but the message always
+                # contains "too long" if we're getting an "MTU exceeded" ICMP response. IF this isn't an "MTU exceeded"
+                # response, then something went wrong, and we can't determine the MTU, so fall back to worst case.
+                return WORST_CASE_MTU_IP
```

Comment:
I don't see the point of `if 'too long' not in e.strerror:` if it does indeed hit it, it will return the same as if it didn't

`else` is only triggered if no exception is raised

which is equivalent to
```
    for attempt in MTU_ATTEMPTS:
        try:
            s.send(b'#' * (attempt - DATAGRAM_HEADER_LENGTH_IN_BYTES))
            return attempt
        except socket.error as e:
            if 'too long' not in e.strerror:
                # We can't rely on the error code, because it varies from platform to platform, but the message always
                # contains "too long" if we're getting an "MTU exceeded" ICMP response. IF this isn't an "MTU exceeded"
                # response, then something went wrong, and we can't determine the MTU, so fall back to worst case.
                pass
    return WORST_CASE_MTU_IP
```

Reply:
I agree it can be simplified, but only marginally. The point of `if 'too long' not in e.strerror:` is that, if we get that error, then something else has gone wrong (perhaps an invalid address or broken network, for example), and we need to just stop looping immediately and return the worst-case. We're likely to just keep getting that error over and over again with every attempt before eventually getting to the worst case. The only error that is acceptable for continuing is the "too long" error. I can, however, get rid of the `else`, move the `return attempt`, and change one `return WORST_CASE_MTU_IP` to a `break`.

```
def _discover_minimum_mtu_to_target(address, port):
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect((address, port))
    s.setsockopt(socket.IPPROTO_IP, IP_MTU_DISCOVER, IP_MTU_DISCOVER_DO)
    for attempt in MTU_ATTEMPTS:
        try:
            s.send(b'#' * (attempt - DATAGRAM_HEADER_LENGTH_IN_BYTES))
            return attempt
        except socket.error as e:
            if 'too long' not in e.strerror:
                # We can't rely on the error code, because it varies from platform to platform, but the message always
                # contains "too long" if we're getting an "MTU exceeded" ICMP response. IF this isn't an "MTU exceeded"
                # response, then something went wrong, and we can't determine the MTU, so fall back to worst case.
                break
            # Now we try again with a lower MTU, which might still be too high
    return WORST_CASE_MTU_IP
```